jamesc4-assign1/crypto.py

```diff
@@ -0,0 +1,362 @@
+    new_alphabet = "XYZ" + string.ascii_uppercase
+    caesar_dict = {}
+
+    # map alphabet characters to those shifted three right, wrapping around
+    for i in range(len(new_alphabet) - 3):
+        caesar_dict[new_alphabet[i]] = new_alphabet[i + 3]
+
+    return caesar_dict
+
+def get_caesar_dict_decrypt():
+    new_alphabet = "XYZ" + string.ascii_uppercase
+    caesar_dict = {}
+
+    # map alphabet characters to those shifted three left, wrapping around
+    for i in range(3, len(new_alphabet)):
+        caesar_dict[new_alphabet[i]] = new_alphabet[i - 3]
+
+    return caesar_dict
+
+def encrypt_caesar(plaintext):
+    """
+    Encrypts plaintext using a Caesar cipher.
+    """
+    letters = ''.join(filter(str.isalpha, plaintext)).upper()
+    caesar_dict = get_caesar_dict_encrypt()
+
+    ciphertext = ""
+
+    # create ciphertext
+    for ch in letters:
+        ciphertext += caesar_dict[ch]
+
+    return ciphertext
+
+
+def decrypt_caesar(ciphertext):
+    """
+    Decrypts a ciphertext using a Caesar cipher.
+    """
+    letters = ''.join(filter(str.isalpha, ciphertext)).upper()
+    caesar_dict = get_caesar_dict_decrypt()
+
+    plaintext = ""
+
+    for ch in letters:
+        plaintext += caesar_dict[ch]
+
+    return plaintext
+
+def format_plaintext_keyword(plaintext, keyword):
+    """
+    Helper for viginere cipher
+    """
+    plaintext = ''.join(filter(str.isalpha, plaintext)).upper()
+    keyword = ''.join(filter(str.isalpha, keyword)).upper()
+    return (plaintext, keyword)    
+
+def get_vigenere_key(keyword, plaintext):
+    """
+    Helper for viginere cipher
+    """
+    key = keyword
+    while len(key) < len(plaintext):
+        key += keyword
+    key = key[:len(plaintext)]
+    return key    
+
+# p, k = 'attack at dawn', '   lemon    '
+# 'LX`OPVE`RNbR'
+
+def encrypt_vigenere(plaintext, keyword):
+    """
+    Encrypts plaintext using a Vigenere cipher with a keyword.
+    """
+    (plaintext, keyword) = format_plaintext_keyword(plaintext, keyword)
+
+    # builds key
+    key = get_vigenere_key(keyword, plaintext)
+
+    ciphertext = ""
+
+    # encrypts the plaintext
+    for i in range(len(key)):
+        ciphertext += chr(ord(plaintext[i]) + ord(key[i]) - ord('A'))
+
+    return ciphertext
+
+
+def decrypt_vigenere(ciphertext, keyword):
+    """
+    Decrypts ciphertext using a Vigenere cipher with a keyword.
+    """
+    (ciphertext, keyword) = format_plaintext_keyword(ciphertext, keyword)
+
+    # builds key
+    key = get_vigenere_key(keyword, ciphertext)
+
+    plaintext = ""
+
+    # decrypts ciphertext
+    for i in range(len(key)):
+        plaintext += chr(ord(ciphertext[i]) - ord(key[i]) + ord('A'))
+
+    return plaintext
+
+
+def encrypt_railfence(plaintext, num_rails):
+    """
+    Encrypts plaintext using a railfence cipher.
+    Performs splicing and ciphertext construction in O(num_rails) time.
+    """
+    letters = ''.join(filter(str.isalpha, plaintext)).upper()
```

Comment:
We actually didn't require you to strip out characters that are not `is alpha` but this is fine!